Ronni-ronsbons/week-8/hw-flask-models/app.py

```diff
@@ -0,0 +1,140 @@
+from flask import Flask, g
+from flask import render_template, flash, redirect, url_for
+
+# import models.py file
+# in Flask and Python, you don't have to export something in order to import it
+# convention that you don't need .py, it's like you're importing a module
+import models
+# import SubForm PostForm, CommentForm classes from forms.py
+from forms import SubForm, PostForm, CommentForm
+
+DEBUG = True
+PORT = 8000
+
+app = Flask(__name__)
+app.secret_key = 'adkjfalj.adflja.dfnasdf.asd'
+
+
+# handle requests when they come in (before)
+# gets snapshot of database before request
+@app.before_request
+def before_request():
+    # g stands for global
+    # access DATABASE variable in models.py file
+    g.db = models.DATABASE
+    g.db.connect()
+
+
+# handle requests when they complete (after)
+# gets response and closes connection to database
+@app.after_request
+def after_request(response):
+    # close connection to database
+    g.db.close()
+    return response
+
+
+# add methods (ex. POST) that root route should accept
+@app.route('/', methods=['GET', 'POST'])
+def index():
+    # code to serve HTML using json data, not needed anymore, now that we are using database
+    # with open('subs.json') as json_data:
+    #     subs = json.load(json_data)
+    #     return render_template('subs.html', subs=subs)
+
+    # return 'Temporary - integrating models'
+
+    # form variable to represent SubForm
+    # [] WHY DO WE NEED PARENTHESES AFTER SUBFORM?
+    form = SubForm()
+    # handle POST request for new sub
+    # check if form submission is valid
+    # [] WHERE DOES .VALIDATE_ON_SUBMIT FUNCTION COME FROM?
```

Comment:
.validadate is a built-in function of the form class. Remember you are extending the form class from flask into a custom for posts. It checks if the form is valid ie does it meet all the required fields.